src/nnfdm.cpp

```diff
@@ -0,0 +1,237 @@
+
+        near_node_flash::data_movement::CreateRequest createRequest(std::string{file}, std::string{dest_file});
+        near_node_flash::data_movement::CreateResponse createResponse;
+
+        near_node_flash::data_movement::RPCStatus status = nnfdm_client->Create(*nnfdm_workflow, createRequest, &createResponse);
+        if (!status.ok()) {
+            axl_abort(-1,
+                "NNFDM Create(%s, %s) failed with error %d (%s) @ %s:%d",
+                file, dest_file, status.error_code(), status.error_message().c_str(), __FILE__, __LINE__
+            );
+            /*NOTREACHED*/
+        }
+
+        switch (createResponse.status()) {
+            case near_node_flash::data_movement::CreateResponse::Status::STATUS_SUCCESS:
+                /* record that the file is in progress */
+                kvtree_util_set_int(elem_hash, AXL_KEY_FILE_STATUS, AXL_STATUS_INPROG);
+
+                /* record the session Id for this file */
+                kvtree_util_set_str(elem_hash, AXL_KEY_FILE_SESSION_UID, createResponse.uid().c_str());
+                break;
+            default:
+                axl_abort(-1,
+                    "NNFDM Offload Create FAILED for %s: error %d (%s) ",
+                    file, createResponse.status(), createResponse.message().c_str()
+                 );
+                 /*NOTREACHED*/
+                 break;
+        } 
+    }
+
+    return AXL_SUCCESS;
+}
+
+int nnfdm_test(int id) {
+    kvtree* file_list = axl_kvtrees[id];
+    kvtree* files     = kvtree_get(file_list, AXL_KEY_FILES);
+
+    /* iterate/wait over in-progress files */
+    for (kvtree_elem* elem = kvtree_elem_first(files); elem != NULL; elem = kvtree_elem_next(elem)) {
+        int status;
+        char* uid;
+
+        kvtree* elem_hash = kvtree_elem_hash(elem);
+        kvtree_util_get_int(elem_hash, AXL_KEY_FILE_STATUS, &status);
+
+        if (status == AXL_STATUS_DEST) {
+            continue;   /* This one is done */
+        }
+
+        kvtree_util_get_str(elem_hash, AXL_KEY_FILE_SESSION_UID, &uid);
+        status = nnfdm_stat(uid, 0);
+
+        if (status != AXL_STATUS_DEST) {
+            return AXL_FAILURE;   /* At least one file is not done */
+        }
+
+        kvtree_util_set_int(elem_hash, AXL_KEY_FILE_STATUS, status);
+    }
+
+    return AXL_SUCCESS;
+}
+
+int nnfdm_wait(int id)
+{
+    const int64_t max_seconds_to_wait{1};
+    kvtree* file_list = axl_kvtrees[id];
+    kvtree* files     = kvtree_get(file_list, AXL_KEY_FILES);
+
+    /* iterate/wait over in-progress files */
+    for (kvtree_elem* elem = kvtree_elem_first(files); elem != NULL; elem = kvtree_elem_next(elem)) {
+        int status;
+        char* uid;
+
+        kvtree* elem_hash = kvtree_elem_hash(elem);
+        kvtree_util_get_str(elem_hash, AXL_KEY_FILE_SESSION_UID, &uid);
+
+        do {
+            status = nnfdm_stat(uid, max_seconds_to_wait);
+        } while (status == AXL_STATUS_INPROG);
```

Comment:
For future work, we may want to add a sleep() in here to avoid thrashing the server will poll requests.